Fix missing Job Control builtins in Bash 3.0-4.3 and stabilize Bash 3.x builds

chaifeng opened the pull request:
This PR fixes missing Job Control commands (`bg`, `disown`, `fg`, `jobs`) in Bash 3.0–4.3 images.

#### Problem
The latest Bash 3.0–4.3 images do not include Job Control builtins. Steps to reproduce with `bash:4.3 (amd64)`:

```sh
docker run --rm bash:4.3@sha256:53abafff8451a86fbb1758af492e5b31740c8f0bc86cf074fbc3d73cbb78e08f -c 'type bg disown fg jobs'
# docker run --rm bash:4.3 -c 'type bg disown fg jobs'
```

Output:
```
bash: line 0: type: bg: not found
bash: line 0: type: disown: not found
bash: line 0: type: fg: not found
bash: line 0: type: jobs: not found
```

The previous image version(amd64) works as expected:

```sh
docker run --rm bash:4.3@sha256:901dc91dfb31a49d5264020556b41bf6e83dd36d879fc745d629d4d3f98210ab -c 'type bg disown fg jobs'
```

Output:
```
bg is a shell builtin
disown is a shell builtin
fg is a shell builtin
jobs is a shell builtin
```

The issue occurs across all Bash 3.0–4.3 images.

Additional Fix

Building Bash 3.x images with parallel jobs causes intermittent failures, such as:
- `ar: savestring.o: No such file or directory`
- `ar: mbutil.o: No such file or directory`
-  ...

Building Bash 3.x with a single thread avoids these failures. This PR adjusts the build process to ensure stability.

Changes
1. Restore Job Control functionality for Bash 3.0–4.3.
2. Disable parallel builds for Bash 3.x to prevent build failures.

chaifeng committed:
Fix missing Job Control builtins in Bash 3.0-4.3 and stabilize Bash 3.x builds

tianon commented:
Just disabling some warnings turned error makes me worry I'll break this again in the future without realizing it.  That combined with these errors being a large part of the patches applied in #43 makes me think perhaps we need more similarly minor patches?

Given this fixes it, is there something in the configure script that needs to be updated to enable job control properly?

chaifeng commented:
Use `awk` to modify the `configure` script for GCC 14 compatibility without extra compiler options.  
- Changed most `main` function declarations in `configure` (except a few).  
- Added `#include <stdlib.h>` to fix warnings for the `exit` method.  

If we don't use a script to modify `configure` and rely on a patch file, the patch will be **very large**.

In my opinion, the Bash 3.0-4.3 source code is quite old. Many of the warnings already existed when compiling with GCC 13. The issue happened because GCC 14 treats warnings as errors by default. This caused the `configure` script to fail and introduced the issue. Using compiler options only makes GCC 14 behave like GCC 13. In this case, I think it’s acceptable, as GCC 13 handled it this way.

tianon commented:
Sorry for the delay; this is a tough one and I've been trying to think about how I want to handle it long-term.  I appreciate what you've done here, and agree that the patch for `configure` is probably less tenable/maintainable than I was hoping. :sob: :heart:

Your `awk` is excellent, but I do worry about my ability to continue to maintain it effectively long-term.  Perhaps your first answer _was_ better, but I don't want to take away from the great work you've put into this. :heart:

I was looking at https://www.gnu.org/software/bash/manual/html_node/Optional-Features.html, hoping something like `./configure --enable-job-control` would let us essentially signal intent to `configure` that we definitely want this enabled and that if it's missing, it should be an error, not that it should quietly swap to disabled, but it turns out that the default _is_ enabled, and so the "disable if not supported" fallback is intentional and complete (even being explicit about enabling it still quietly disables it again).

In other words, the script doesn't differentiate between "auto" and "yes, please" even in the latest revision. :sob:

So, I guess long term we probably need to come up with explicit tests for the features we care about supporting so this doesn't happen again in the future. :disappointed:

To get this in, I'm going to update this PR to revert c7f927dd1c89186d35621313964eaf3dba9e6742, but leaving the explicit commit reference in this comment so I can consider coming back to it in the future if this gets harder on the next round (although at some point I'll probably have to consider no longer building/updating these instead). :bow:

I'll also try to think about how to write some better integration tests for this kind of stuff in the future -- for this specific one, I guess I could just do something basic like `command -v jobs`, but it might be worth doing something slightly more?
